flytekit/annotated/context_manager.py

```diff
@@ -0,0 +1,259 @@
+import os
+from contextlib import contextmanager
+from typing import List, Optional, Generator, Dict, Any
+import datetime as _datetime
+import logging as _logging
+from enum import Enum
+
+from flytekit import __version__ as _api_version
+from flytekit.clients import friendly as flyte_client
+from flytekit.common import constants as _constants
+from flytekit.common.exceptions import user as _user_exception
+from flytekit.common.nodes import SdkNode
+from flytekit.common.tasks.sdk_runnable import ExecutionParameters
+from flytekit.configuration import sdk as _sdk_config, internal as _internal_config
+from flytekit.interfaces.data import common as _common_data
+from flytekit.interfaces.data.gcs import gcs_proxy as _gcs_proxy
+from flytekit.interfaces.data.http import http_data_proxy as _http_data_proxy
+from flytekit.interfaces.data.local import local_file_proxy as _local_file_proxy
+from flytekit.interfaces.data.s3 import s3proxy as _s3proxy
+from flytekit.models.core import identifier as _identifier
+from flytekit.interfaces.stats.taggable import get_stats as _get_stats
+from flytekit.engines.unit import mock_stats as _mock_stats
+
+
+class CompilationState(object):
+    def __init__(self):
+        self.nodes: List[SdkNode] = []
+        self.mode = 1  # TODO: Turn into enum in the future, or remove if only one mode.
+
+
+class ExecutionState(object):
```

Comment:
i am really liking where we are heading

Reply:
i just think the context stack should be separate from the actual context - composition this way will make unit test easier.

So create ExecutionContext
ExecutionContextStack?

Reply:
actually i mean this comment for FlyteContext

Reply:
I agree, it's awkward - can you think of a way around it?